### PR TITLE
Make metad works stable on multi replicas cluster

### PR DESCRIPTION
**Pull request by dangleptr**
Main changes:
1.  Refactor clusterId and ActiveHostMan
2. make FLAGS_hosts_whitelist_enabled to be false by default,  so we don't need add host anymore.
3.  Fix some bugs.

### CODE REVIEW

**Comment by dangleptr** [comment]
Jenkins go

File: src/daemons/MetaDaemon.cpp [hunk 1]
```diff
@@ -69,9 +76,50 @@ std::unique_ptr<nebula::kvstore::KVStore> initKV(std::vector<nebula::HostAddr> p
...
     if (!(kvstore->init())) {
-        LOG(ERROR) << "nebula store init failed";
+        LOG(ERROR) << "Nebula store init failed";
         return nullptr;
     }
+
+    LOG(INFO) << "Waiting for the leader elected...";
+    nebula::HostAddr leader;
+    while (true) {
+        auto ret = kvstore->partLeader(0, 0);
```
Comment (wadeliuyi) [inline src/daemons/MetaDaemon.cpp]: what mean about (0, 0)?

Comment (dangleptr) [reply src/daemons/MetaDaemon.cpp]: For meta,  we only have one space (0) and one part (0)

File: src/meta/client/MetaClient.cpp [hunk 2]
```diff
@@ -999,27 +1008,26 @@ StatusOr<SchemaVer> MetaClient::getNewestEdgeVerFromCache(const GraphSpaceID& sp
 
 
 folly::Future<StatusOr<bool>> MetaClient::heartbeat() {
+    if (clusterId_.load() == 0) {
+        clusterId_ = ClusterIdMan::getClusterIdFromFile(FLAGS_cluster_id_path);
```
Comment (critical27) [inline src/meta/client/MetaClient.cpp]: clusterId_.store()?

File: src/meta/client/MetaClient.cpp [hunk 3]
```diff
@@ -999,27 +1008,26 @@ StatusOr<SchemaVer> MetaClient::getNewestEdgeVerFromCache(const GraphSpaceID& sp
...
-            && !clusterMan_->isClusterIdSet()) {
-            clusterMan_->setClusterId(resp.get_clusterId());
-            if (!clusterMan_->clusterIdDumped()) {
-                if (!clusterMan_->dumpClusterId()) {
-                    LOG(ERROR) << "meta client clusterId dump failed!";
-                    return false;
-                }
+        if (clusterId_.load() == 0) {
+            LOG(INFO) << "Persisit the cluster Id from metad " << resp.get_cluster_id();
+            if (ClusterIdMan::persistInFile(resp.get_cluster_id(), FLAGS_cluster_id_path)) {
```
Comment (critical27) [inline src/meta/client/MetaClient.cpp]: Do we need return false when persist failed?

**Comment by sherman-the-tank** [comment]
Again, I think we need more discussion on using clusterId to prevent unexpected hosts joining the cluster. I'll suggest to hold on a little so that we have enough time to think through all scenarios

Let's have a quick offline discussion

File: src/daemons/MetaDaemon.cpp (hunk 1 above)
Comment (wadeliuyi) [reply src/daemons/MetaDaemon.cpp]: maybe we can replace this magic number to some readable word, that coder understand it more easy.

Comment (sherman-the-tank) [reply src/daemons/MetaDaemon.cpp]: Or make the previous reply as a comment in front of this line

File: src/common/base/Status.h [hunk 4]
```diff
@@ -60,6 +60,10 @@ class Status final {
         return code() == rhs.code();
     }
 
+    bool operator!=(const Status &rhs) const {
```
Comment (sherman-the-tank) [inline src/common/base/Status.h]: Why do we need this? My understanding is the compiler will convert ```!=``` to ```! ==()```

File: src/daemons/MetaDaemon.cpp [hunk 5]
```diff
@@ -30,9 +30,9 @@ using nebula::Status;
 DEFINE_int32(port, 45500, "Meta daemon listening port");
 DEFINE_bool(reuse_port, true, "Whether to turn on the SO_REUSEPORT option");
 DEFINE_string(data_path, "", "Root data path");
-DEFINE_string(peers, "", "It is a list of IPs split by comma,"
-                         "the ips number equals replica number."
-                         "If empty, it means replica is 1");
+DEFINE_string(meta_server_addrs, "", "It is a list of IPs split by comma,"
+                                     "the ips number equals replica number."
```
Comment (sherman-the-tank) [inline src/daemons/MetaDaemon.cpp]: This line conflicts with the next sentence. If the number of IPs equals to the number of replicas, it means the IP list includes the local IP as well. So when the IP list is empty, the number of replica should be zero

File: src/daemons/MetaDaemon.cpp (hunk 1 above)
Comment (sherman-the-tank) [reply src/daemons/MetaDaemon.cpp]: > For meta, we only have one space (0) and one part (0)

Put this as a comment in the code?

File: src/meta/ActiveHostsMan.cpp [hunk 6]
```diff
@@ -6,150 +6,52 @@
...
-            updateHostInfo({host.ip, host.port}, info);
+        HostInfo info = HostInfo::decode(iter->val());
+        if (now - info.lastHBTimeInSec_ < expiredTTL) {
+            hosts.emplace_back(host.ip, host.port);
         }
         iter->next();
     }
-}
-
-void ActiveHostsMan::cleanExpiredHosts() {
```
Comment (sherman-the-tank) [inline src/meta/ActiveHostsMan.cpp]: We still need this to remove the ghost hosts that are dead for too long, don't we?

File: src/daemons/MetaDaemon.cpp (hunk 1 above)
Comment (dangleptr) [reply src/daemons/MetaDaemon.cpp]: Good suggestion.  I will fix it.

File: src/common/base/Status.h (hunk 4 above)
Comment (dangleptr) [reply src/common/base/Status.h]: No.  The compiler will not do that.

File: src/meta/ActiveHostsMan.cpp (hunk 6 above)
Comment (dangleptr) [reply src/meta/ActiveHostsMan.cpp]: Yes.  I am planing to do it in compaction.

File: src/meta/client/MetaClient.cpp (hunk 2 above)
Comment (dangleptr) [reply src/meta/client/MetaClient.cpp]: They are the same thing.

File: src/meta/client/MetaClient.cpp (hunk 3 above)
Comment (dangleptr) [reply src/meta/client/MetaClient.cpp]: Good point.

File: src/daemons/MetaDaemon.cpp (hunk 5 above)
Comment (dangleptr) [reply src/daemons/MetaDaemon.cpp]: Good catch.

**Comment by dangleptr** [comment]
#736

**Pull request merged**